client/src/send_and_confirm_transactions_in_parallel.rs

```diff
@@ -0,0 +1,417 @@
+                    .iter()
+                    .filter(|x| {
+                        // all transactions that are not expired
+                        let is_not_expired = current_blockheight <= x.last_valid_blockheight;
+                        // all transaction that expired between last and current check
+                        let is_recently_expired = last_block_height <= x.last_valid_blockheight
+                            && current_blockheight > x.last_valid_blockheight;
+                        is_not_expired || is_recently_expired
+                    })
+                    .map(|x| *x.key())
+                    .collect();
+                for signatures in
+                    transactions_to_verify.chunks(MAX_GET_SIGNATURE_STATUSES_QUERY_ITEMS)
+                {
+                    if let Ok(result) = rpc_client.get_signature_statuses(signatures).await {
+                        let statuses = result.value;
+                        for (signature, status) in signatures.iter().zip(statuses.into_iter()) {
+                            if let Some(status) = status {
+                                if status.satisfies_commitment(rpc_client.commitment()) {
+                                    if let Some((_, data)) = transaction_map.remove(signature) {
+                                        confirmed_transactions.fetch_add(1, Ordering::Relaxed);
+                                        transaction_errors[data.index] = status.err;
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+
+                last_block_height = current_blockheight;
+            }
+            let elapsed = instant.elapsed();
+            if elapsed < CONFIRMATION_REFRESH_RATE {
+                tokio::time::sleep(CONFIRMATION_REFRESH_RATE - elapsed).await;
+            }
+        }
+        transaction_errors
+    })
+}
+
+async fn sign_all_messages_and_send<T: Signers + ?Sized>(
+    progress_bar: &Option<indicatif::ProgressBar>,
+    tpu_client: &TpuClient<QuicPool, QuicConnectionManager, QuicConfig>,
+    transaction_map: Arc<DashMap<Signature, TransactionData>>,
+    messages_with_index: Vec<(usize, Message)>,
+    blockhash_data_rw: Arc<RwLock<BlockHashData>>,
+    signers: &T,
+    confirmed_transactions: Arc<AtomicU32>,
+    total_transactions: usize,
+) -> Result<()> {
+    let current_transaction_count = messages_with_index.len();
+    // send all the transaction meesages
+    for (counter, (index, message)) in messages_with_index.iter().enumerate() {
+        let mut transaction = Transaction::new_unsigned(message.clone());
+        let blockhashdata = *blockhash_data_rw.read().await;
+        transaction.try_sign(signers, blockhashdata.blockhash)?;
+        let signature = transaction.signatures[0];
+        if !tpu_client.send_transaction(&transaction).await {
```

Comment:
we should probably split `sign` and `send` to separate loops to avoid abandoning previously broadcasted transactions should `sign` fail somehow. also consider implications of the early-return